lib/crypto.js

```diff
@@ -95,6 +96,25 @@ Hmac.prototype.digest = Hash.prototype.digest;
 Hmac.prototype._flush = Hash.prototype._flush;
 Hmac.prototype._transform = Hash.prototype._transform;
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+//
+// Note: hmac object can not be used after validate() method has been called.
+Hmac.prototype.validate = function(inputBuffer) {
+  var ah = new Hmac('sha256', this.key).update(this.digest()).digest();
```

Comment:
I am a bit worried about usage of `sha256` here. Could this possibly adapt to the digest size specified in `Hmac` constructor?

Like, for `sha512` hmac it could use `sha512` verifier hmacs too. Does it make sense?